src/NHibernate/Linq/Visitors/ExpressionParameterVisitor.cs

```diff
@@ -125,11 +116,14 @@ protected override Expression VisitConstant(ConstantExpression expression)
 
 				// We have a bit more information about the null parameter value.
 				// Figure out a type so that HQL doesn't break on the null. (Related to NH-2430)
+				// In v5.3 types are calculated by ConstantTypeLocator, this logic is only for back compatibility.
+				// TODO 6.0: Remove
 				if (expression.Value == null)
 					type = NHibernateUtil.GuessType(expression.Type);
 
 				// Constant characters should be sent as strings
-				if (expression.Type == typeof(char))
+				// TODO 6.0: Remove
+				if (_queryVariables == null && expression.Type == typeof(char))
 				{
 					value = value.ToString();
```

Comment:
This to me looks as a workaroud to fix `CharIndexFunction` and `CharIndexOffsetNegativeFunction` tests on Sql Server and Sql CE databases. Instead of converting it to string, this PR set the correct parameter size for drivers that fail without it.